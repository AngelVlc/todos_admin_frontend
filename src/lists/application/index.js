export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
export * from "./GetListsUseCase";
export * from "./CreateListUseCase";
export * from "./UpdateListUseCase";
export * from "./MoveListItemUseCase";
export * from "./IndexAllListsUseCase";
export * from "./GetSearchSecureKeyUseCase";
