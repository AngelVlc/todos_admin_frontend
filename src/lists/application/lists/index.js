export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
