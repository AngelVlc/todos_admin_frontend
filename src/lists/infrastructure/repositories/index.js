export * from "./ListsRepository";
export * from "./ListItemsRepository";
