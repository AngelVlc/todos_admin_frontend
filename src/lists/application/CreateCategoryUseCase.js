import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class CreateCategoryUseCase extends BaseUseCase {
  _categoriesRepository;

  constructor({ categoriesRepository }) {
    super();

    this._categoriesRepository = categoriesRepository;
  }

  static create() {
    const categoriesRepository = new CategoriesRepository();

    return new CreateCategoryUseCase({ categoriesRepository });
  }

  async execute(category) {
   return await this._categoriesRepository.create(category);
  }
}
