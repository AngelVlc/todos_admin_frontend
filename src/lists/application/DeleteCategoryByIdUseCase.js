import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class DeleteCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new DeleteCategoryByIdUseCase({ repository });
  }

  async execute(categoryId) {
    return await this._repository.deleteById(categoryId);
  }
}
