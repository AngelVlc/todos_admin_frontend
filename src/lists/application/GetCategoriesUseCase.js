import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoriesUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoriesUseCase({ repository });
  }

  async execute() {
    const result = await this._repository.getAll();

    return result.map((item) => {
      return new Category(item);
    });
  }
}
