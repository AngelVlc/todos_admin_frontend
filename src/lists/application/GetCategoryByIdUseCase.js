import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoryByIdUseCase({ repository });
  }

  async execute(id) {
    const result = await this._repository.getById(id);

    return new Category(result);
  }
}
