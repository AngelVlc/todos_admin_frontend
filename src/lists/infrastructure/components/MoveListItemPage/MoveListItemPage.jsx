import React, { useEffect, useState, useContext, useCallback } from "react";
import { useParams, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";
import {
  GetListByIdUseCase,
  GetListsUseCase,
  MoveListItemUseCase,
} from "../../../application/lists";
import "./MoveListItemForm.css";

export const MoveListItemPage = () => {
  const [list, setList] = useState();
  const [lists, setLists] = useState();
  const [itemTitle, setItemTitle] = useState();
  const [selectedDestination, setSelectedDestination] = useState(null);
  const { useCaseFactory } = useContext(AppContext);
  let { listId, listItemId } = useParams();
  listId = parseInt(listId);
  listItemId = parseInt(listItemId);
  let history = useHistory();

  const getList = useCallback(async () => {
    const getListByIdUseCase = useCaseFactory.get(GetListByIdUseCase);
    const list = await getListByIdUseCase.execute(listId);

    setList(list);

    for (const item of list.items) {
      if (item.id === listItemId) {
        setItemTitle(item.title);
      }
    }
  }, [listId, listItemId, useCaseFactory]);

  const getLists = useCallback(async () => {
    const getListsUseCase = useCaseFactory.get(GetListsUseCase);
    const data = await getListsUseCase.execute();

    setLists(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getList();
    getLists();
  }, [listId, getList, getLists]);

  const moveListItem = async () => {
    if (!selectedDestination) {
      return;
    }

    const useCase = useCaseFactory.get(MoveListItemUseCase);
    const result = await useCase.execute(listId, listItemId, selectedDestination);

    if (result) {
      history.push(`/lists/${listId}`);
    }

    console.log(selectedDestination);
  };

  return (
    <>
      {list && (
        <div className="container">
          <Breadcrumb
            items={[
              { url: "/lists", text: "Lists" },
              { url: `/lists/${listId}/edit`, text: list.name },
              { url: `/lists/${listId}/moveItem`, text: "Move Item" },
            ]}
          />
          <h3 className="title">{`Move list item ${itemTitle}`}</h3>

          <div className="control radio-list ml-4 mb-5">
            <>
              {lists &&
                lists.length > 0 &&
                lists
                  .filter((item) => item.id !== listId)
                  .map((item, index) => (
                    <div key={index} className="mb-3">
                      <label className="radio">
                        <input
                          type="radio"
                          name="destination"
                          className="mr-2"
                          value={item.id}
                          onChange={() => setSelectedDestination(item.id)}
                          data-testid={`radio-${item.id}`}
                        />
                        {item.name} ({item.itemsCount} items)
                      </label>
                    </div>
                  ))}
            </>
          </div>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="yes"
              onClick={() => moveListItem()}
            >
              YES
            </button>
            <button
              className="button"
              data-testid="no"
              onClick={() => history.goBack()}
            >
              NO
            </button>
          </div>
        </div>
      )}
    </>
  );
};
