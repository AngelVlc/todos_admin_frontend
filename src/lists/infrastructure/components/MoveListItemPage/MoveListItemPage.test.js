import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { MoveListItemPage } from "./MoveListItemPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { List } from "../../../domain";
import { GetListByIdUseCase, GetListsUseCase, MoveListItemUseCase } from "../../../application/lists";

const mockHistoryGoBack = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    goBack: mockHistoryGoBack,
    push: mockHistoryPush,
  }),
}));

const mockedGetListByIdUseCase = {
  execute: jest.fn(),
};

const mockedGetListsUseCase = {
  execute: jest.fn(),
};

const mockedMoveListItemUseCase = {
  execute: jest.fn(),
};

const renderWithContextAndRouter = () => {
  mockedGetListByIdUseCase.execute.mockResolvedValue(
    new List({ id: 2, name: "ListName", items: [{ id: 5, title: "title" }] })
  );
  mockedGetListsUseCase.execute.mockResolvedValue([
    new List({ id: 20, name: "Another list", itemsCount: 2 }),
    new List({ id: 30, name: "Another more", itemsCount: 3 }),
  ]);

  const useCaseFactory = {
    get: (useCase) => {
      switch (useCase) {
        case GetListByIdUseCase:
          return mockedGetListByIdUseCase;
        case GetListsUseCase:
          return mockedGetListsUseCase;
        case MoveListItemUseCase:
          return mockedMoveListItemUseCase
      }
    },
  };

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/lists/2/move-item/5`]}>
          <Route path="/lists/:listId/move-item/:listItemId">{<MoveListItemPage />}</Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

describe("MoveListItemPage", () => {
  it("should match the snapshot", async () => {
    let fragment;
    await act(async () => {
      const { asFragment } = renderWithContextAndRouter();
      fragment = asFragment;
    });

    expect(fragment()).toMatchSnapshot();
  });

  it("when click on cancel should cancel the deletion", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("no"));
    });

    expect(mockHistoryGoBack).toHaveBeenCalled();
    mockHistoryGoBack.mockClear();
  });

  it("when click on yes should move the list item", async () => {
    const container = renderWithContextAndRouter();

    mockedMoveListItemUseCase.execute.mockResolvedValue(true);

    await waitFor(() => {
      fireEvent.click(container.getByTestId("radio-30"));
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("yes"));
    });

    expect(mockedMoveListItemUseCase.execute).toHaveBeenCalled();
    console.log("#######", mockedMoveListItemUseCase.execute.mock.calls[0])
    expect(mockedMoveListItemUseCase.execute.mock.calls[0][0]).toBe(2);
    expect(mockedMoveListItemUseCase.execute.mock.calls[0][1]).toBe(5);
    expect(mockedMoveListItemUseCase.execute.mock.calls[0][2]).toBe(30);
    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
    mockHistoryPush.mockClear();
  });
});