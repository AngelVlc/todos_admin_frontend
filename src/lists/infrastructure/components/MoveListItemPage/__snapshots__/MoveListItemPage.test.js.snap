// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveListItemPage should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/lists"
          >
            Lists
          </a>
        </li>
        <li>
          <a
            href="/lists/2/edit"
          >
            ListName
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists/2/moveItem"
          >
            Move Item
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      Move list item title
    </h3>
    <div
      class="control radio-list ml-4 mb-5"
    >
      <div
        class="mb-3"
      >
        <label
          class="radio"
        >
          <input
            class="mr-2"
            data-testid="radio-20"
            name="destination"
            type="radio"
            value="20"
          />
          Another list (2 items)
        </label>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="radio"
        >
          <input
            class="mr-2"
            data-testid="radio-30"
            name="destination"
            type="radio"
            value="30"
          />
          Another more (3 items)
        </label>
      </div>
    </div>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;
